backends/p4tools/modules/testgen/targets/bmv2/test/testgen_api/control_plane_filter_test.cpp

```diff
@@ -0,0 +1,264 @@
+    templateString << R"p4(
+header ethernet_t {
+    bit<48> dst_addr;
+    bit<48> src_addr;
+    bit<16> ether_type;
+}
+
+struct Headers {
+  ethernet_t eth_hdr;
+}
+
+struct Metadata {  }
+parser parse(packet_in pkt, out Headers hdr, inout Metadata m, inout standard_metadata_t sm) {
+  state start {
+      pkt.extract(hdr.eth_hdr);
+      transition accept;
+  }
+}
+control ingress(inout Headers hdr, inout Metadata meta, inout standard_metadata_t sm) {)p4"
+                   << ingressBlock << R"p4(
+}
+control egress(inout Headers hdr, inout Metadata meta, inout standard_metadata_t sm) {
+  apply {}
+}
+control deparse(packet_out pkt, in Headers hdr) {
+  apply {
+    pkt.emit(hdr.eth_hdr);
+  }
+}
+control verifyChecksum(inout Headers hdr, inout Metadata meta) {
+  apply {}
+}
+control computeChecksum(inout Headers hdr, inout Metadata meta) {
+  apply {}
+}
+V1Switch(parse(), verifyChecksum(), ingress(), egress(), computeChecksum(), deparse()) main;
+)p4";
+    return P4_SOURCE(P4Headers::V1MODEL, templateString.str().c_str());
+}
+
+CompilerOptions generateDefaultApiTestCompilerOptions() {
+    auto compilerOptions = CompilerOptions();
+    compilerOptions.target = "bmv2";
+    compilerOptions.arch = "v1model";
+    return compilerOptions;
+}
+
+P4Tools::P4Testgen::TestgenOptions &generateDefaultApiTestTestgenOptions() {
+    auto &testgenOptions = P4Tools::P4Testgen::TestgenOptions::get();
+    testgenOptions.testBackend = "PROTOBUF_IR";
+    testgenOptions.testBaseName = "dummy";
+    testgenOptions.seed = 1;
+    testgenOptions.maxTests = 0;
+    // Create a bespoke packet for the Ethernet extract call.
+    constexpr int kEthHdrBitSize = 112;
+    testgenOptions.minPktSize = kEthHdrBitSize;
+    testgenOptions.maxPktSize = kEthHdrBitSize;
+    return testgenOptions;
+}
+
+TEST(P4TestgenControlPlaneFilterTest, GeneratesCorrectTests) {
+    auto source = generateTestProgram(R"(
+    // Drop the packet.
+    action acl_drop() {
+        mark_to_drop(sm);
+    }
+
+    table drop_table {
+        key = {
+            hdr.eth_hdr.dst_addr : ternary @name("dst_eth");
+        }
+        actions = {
+            acl_drop();
+            @defaultonly NoAction();
+        }
+    }
+
+    apply {
+        if (hdr.eth_hdr.isValid()) {
+            drop_table.apply();
+        }
+    })");
+    auto compilerOptions = generateDefaultApiTestCompilerOptions();
+    auto &testgenOptions = generateDefaultApiTestTestgenOptions();
+
+    // First, we ensure that tests are generated correctly. We expect two tests.
+    // One which exercises action acl_drop and one which exercises the default action, NoAction.
+    auto testListOpt =
+        P4Tools::P4Testgen::Testgen::generateTests(source, compilerOptions, testgenOptions);
+
+    ASSERT_TRUE(testListOpt.has_value());
+    auto testList = testListOpt.value();
+    ASSERT_EQ(testList.size(), 2);
+    const auto *testWithDrop = testList[0]->checkedTo<P4Tools::P4Testgen::Bmv2::ProtobufIrTest>();
```

Comment:
Should it be clear why you always expect the drop one to be first? Seems like an implementation detail? Probably better to loop through them and check that at least one is a drop and at least one exercises NoAction?

Reply:
Tried to implement something here. Any way to use a lambda in Protobuf matchers? Currently I need to convert the entire vector to strings before passing them to the matchers.